src/verification.js

```diff
@@ -59,41 +59,82 @@ function UntrustedDeviceDialog(props) {
         </div>
     </BaseDialog>;
 }
-
+/**
+ * This will verify an individual device in a way
+ * that is compatible with non-cross-singing clients
+ * (sending a m.key.verification.start event straight away),
+ * including older riots.
+ *
+ * If the verification succeeds,
+ * the trust will be local to this session,
+ * and not part of cross-signing.
+ *
+ * Only SAS and comparing fingerprints is supported.
+ *
+ * @param  {User | Member} user the user
+ * @param  {DeviceInfo} device the device
+ * @return {Promise} resolves when the verification has finished
+ */
 export async function verifyDevice(user, device) {
     if (!await enable4SIfNeeded()) {
         return;
     }
-    Modal.createTrackedDialog("Verification warning", "unverified session", UntrustedDeviceDialog, {
-        user,
-        device,
-        onFinished: async (action) => {
-            if (action === "sas") {
-                const cli = MatrixClientPeg.get();
-                const verificationRequestPromise = cli.legacyDeviceVerification(
-                    user.userId,
-                    device.deviceId,
-                    verificationMethods.SAS,
-                );
-                dis.dispatch({
-                    action: "set_right_panel_phase",
-                    phase: RIGHT_PANEL_PHASES.EncryptionPanel,
-                    refireParams: {member: user, verificationRequestPromise},
-                });
-            } else if (action === "legacy") {
-                const ManualDeviceKeyVerificationDialog = sdk.getComponent("dialogs.ManualDeviceKeyVerificationDialog");
-                Modal.createTrackedDialog("Legacy verify session", "legacy verify session",
-                    ManualDeviceKeyVerificationDialog,
-                    {
-                        userId: user.userId,
-                        device,
-                    },
-                );
-            }
-        },
+    return new Promise(resolve => {
```

Comment:
this could be massively cleaned up by using 

```
const {finished} = Modal.create...
const [action] = await finished;
```

Reply:
noted

Reply:
The dialog gets closed before the verification is finished though...

Reply:
Right, I just mean in terms of flattening the code